main.go

```diff
@@ -0,0 +1,482 @@
+	var nodes corev1.NodeList
+	if err := client.List(context.Background(), k8s.AllNamespaces, &nodes); err != nil {
+		log.Fatal().Err(err).Msg("Could not retrieve the list of nodes.")
+	}
+
+	var allPods corev1.PodList
+
+	if err := client.List(context.Background(), k8s.AllNamespaces, &allPods); err != nil {
+		log.Fatal().Err(err).Msg("Could not retrieve the list of pods.")
+	}
+
+	nodesByPool := groupNodesByPool(nodes.Items)
+
+	for pool, nodes := range nodesByPool {
+		log.Info().Msgf("Node pool: %s\n", pool)
+
+		nodeInfos := collectNodeInfos(nodes, allPods.Items)
+		reportNodePoolMetrics(pool, nodeInfos)
+
+		nodeCountUnderLimit := 0
+		nodeCountScaleDownInProgress := 0
+
+		// For every node pool we check if there are enough nodes using less resources than the limit for scaledown.
+		for _, nodeInfo := range nodeInfos {
+			nodeLabels := nodeInfo.labels
+
+			if isNodeUnderutilizedCandidate(nodeInfo) {
+				nodeCountUnderLimit++
+			}
+
+			if nodeLabels.scaleDownInProgress {
+				nodeCountScaleDownInProgress++
+			}
+		}
+
+		log.Info().Msgf("Number of underutilized nodes: %d", nodeCountUnderLimit)
+		log.Info().Msgf("Number of nodes already being removed: %d", nodeCountScaleDownInProgress)
+
+		// We check if there are enough underutilized pods so that we can initiate a scaledown.
+		// NOTE: We multiply by (nodeCountScaleDownInProgress + 1), because there might be nodes for which
+		// we have initiated the scaledown in previous iterations already, which haven't been removed yet,
+		// and we have to take these into account.
+		if nodeCountUnderLimit >= nodeInfos[0].labels.scaleDownRequiredUnderutilizedNodeCount*(nodeCountScaleDownInProgress+1) {
+			pick := pickUnderutilizedNodeToRemove(nodeInfos)
+
+			if pick == nil {
+				log.Info().Msg("No node was picked for removal.")
+			} else {
+				log.Info().Msg("The node picked for removal:")
+				log.Info().Msgf("Node %v\n", *pick.node.Metadata.Name)
+				log.Info().Msgf("Allocatable CPU: %vm, memory: %vMi\n", pick.stats.allocatableCPU, pick.stats.allocatableMemoryMB)
+				log.Info().Msgf("Pods on node total requests, CPU: %vm, memory: %vMi\n", pick.stats.totalCPURequests, pick.stats.totalMemoryRequests)
+				log.Info().Msgf("CPU utilization: %v%%, memory utilization: %v%%\n", pick.stats.utilizedCPURatio*100, pick.stats.utilizedMemoryRatio*100)
+
+				log.Info().Msg("Cordoning the node...")
+				cordonAndMarkNode(pick.node, client)
+
+				log.Info().Msg("Drain the pods...")
+				drainPods(pick, client)
+			}
+		}
+	}
+}
+
+func cordonAndMarkNode(node *corev1.Node, k8sClient *k8s.Client) error {
+	*node.Spec.Unschedulable = true
+
+	// We add an explicit label so in the next iteration we know that this node has already been picked for removal.
+	node.Metadata.Labels[labelNodeCompactorScaleDownInProgress] = "true"
+
+	err := k8sClient.Update(context.Background(), node)
+
+	return err
+}
+
+func drainPods(node *nodeInfo, k8sClient *k8s.Client) error {
```

Comment:
Is this draining process the same as [kubectl drain](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#drain)?

Reply:
I think `kubectl drain` is more sophisticated, but pretty much it also just deletes all the pods from the node.

Reply:
It would be good to check https://github.com/estafette/estafette-gke-preemptible-killer/blob/cf013fb243ab559042892406c6ec7a2671fac54d/kubernetes.go#L162, since it's a bit more advanced by filtering out daemonsets and kube-system components from getting deleted.

Reply:
I implemented the filtering of DaemonSets and `kube-system` pods.

I haven't implemented the waiting for the drain to be done, because for the compactor it's not really relevant. It would be important if the draining fails and then the node is stuck in an Unschedulable state. I want to start testing the compactor and see if that ever happens.